// @cloudbase/framework-core/lib/plugin-sevice-api/index.js
/**
 *
 * Copyright 2020 Tencent
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
const link_1 = require('../utils/link');
const generator_1 = require('../generator');
const api_1 = require('../api');
const emoji_1 = require('../utils/emoji');
/**
 * 插件服务注入为插件提供的 API
 */
class PluginServiceApi {
  constructor(pluginManager) {
    this.pluginManager = pluginManager;
  }
  /**
     * 获取 manager 实例
     */
  get cloudbaseManager() {
    return this.pluginManager.context.cloudbaseManager;
  }
  /**
     * 获取项目根路径
     */
  get projectPath() {
    return this.pluginManager.context.projectPath;
  }
  /**
     * 项目环境id
     */
  get envId() {
    return this.pluginManager.context.envId;
  }
  /**
     * secretId
     */
  get secretId() {
    return this.pluginManager.context.cloudbaseConfig.secretId || '';
  }
  /**
     * secretKey
     */
  get secretKey() {
    return this.pluginManager.context.cloudbaseConfig.secretKey || '';
  }
  /**
     * token
     */
  get token() {
    return this.pluginManager.context.cloudbaseConfig.token || '';
  }
  /**
     * 获取 logger 实例
     */
  get logger() {
    return this.pluginManager.context.logger;
  }
  /**
     * 获取资源操作 API
     */
  get resourceProviders() {
    return this.pluginManager.context.resourceProviders;
  }
  /**
     * 获取云开发 CLI 配置文件
     */
  get projectConfig() {
    return this.pluginManager.context.projectConfig;
  }
  /**
     * 获取generator实例
     */
  get generator() {
    return new generator_1.Generator();
  }
  /** 云API */
  get cloudApi() {
    return api_1.CloudApi;
  }
  /**
     * 生成命令行链接
     */
  get genClickableLink() {
    return link_1.genClickableLink;
  }
  /**
     * 安全地使用emoji
     */
  get emoji() {
    return emoji_1.emoji;
  }
}
exports.default = PluginServiceApi;
